src/core/lib/surface/server.cc

```diff
@@ -1435,30 +1452,40 @@ static grpc_call_error queue_call_request(grpc_server* server, size_t cq_idx,
       rm = &rc->data.registered.method->matcher;
       break;
   }
-  if (gpr_locked_mpscq_push(&rm->requests_per_cq[cq_idx], &rc->request_link)) {
-    /* this was the first queued request: we need to lock and start
-       matching calls */
-    gpr_mu_lock(&server->mu_call);
-    while ((calld = rm->pending_head) != nullptr) {
-      rc = reinterpret_cast<requested_call*>(
-          gpr_locked_mpscq_pop(&rm->requests_per_cq[cq_idx]));
-      if (rc == nullptr) break;
-      rm->pending_head = calld->pending_next;
-      gpr_mu_unlock(&server->mu_call);
-      if (!gpr_atm_full_cas(&calld->state, PENDING, ACTIVATED)) {
-        // Zombied Call
-        GRPC_CLOSURE_INIT(
-            &calld->kill_zombie_closure, kill_zombie,
-            grpc_call_stack_element(grpc_call_get_call_stack(calld->call), 0),
-            grpc_schedule_on_exec_ctx);
-        GRPC_CLOSURE_SCHED(&calld->kill_zombie_closure, GRPC_ERROR_NONE);
-      } else {
-        publish_call(server, calld, cq_idx, rc);
-      }
-      gpr_mu_lock(&server->mu_call);
-    }
+
+  // Fast path: if this is not the first request or there is no pending request
+  //            to be processed, immediately return.
+  if (!gpr_locked_mpscq_push(&rm->requests_per_cq[cq_idx], &rc->request_link) ||
+      // Note: We are reading the process_pending_head without holding the
+      //       server's call mutex. We need to gaurantee that this load is not
+      //       reordered with the store in publish_new_rpc(). Otherwise, we may
+      //       read a false value and miss processing the pending head.
+      !rm->process_pending_head.load(std::memory_order_acquire)) {
+    return GRPC_CALL_OK;
+  }
+  // Slow path: This was the first queued request and there are pendings:
+  //            We need to lock and start matching calls.
+  gpr_mu_lock(&server->mu_call);
+  while ((calld = rm->pending_head) != nullptr) {
+    rc = reinterpret_cast<requested_call*>(
+        gpr_locked_mpscq_pop(&rm->requests_per_cq[cq_idx]));
+    if (rc == nullptr) break;
+    rm->pending_head = calld->pending_next;
     gpr_mu_unlock(&server->mu_call);
+    if (!gpr_atm_full_cas(&calld->state, PENDING, ACTIVATED)) {
+      // Zombied Call
+      GRPC_CLOSURE_INIT(
+          &calld->kill_zombie_closure, kill_zombie,
+          grpc_call_stack_element(grpc_call_get_call_stack(calld->call), 0),
+          grpc_schedule_on_exec_ctx);
+      GRPC_CLOSURE_SCHED(&calld->kill_zombie_closure, GRPC_ERROR_NONE);
+    } else {
+      publish_call(server, calld, cq_idx, rc);
+    }
+    gpr_mu_lock(&server->mu_call);
   }
+  rm->process_pending_head.store(false, std::memory_order_relaxed);
```

Comment:
Thank you. That's a great question. I thought of that scenario and could explain it as infeasible, because Thread A at Step 3, will see a non empty queue and won't add to the pending_head:
https://github.com/grpc/grpc/blob/master/src/core/lib/surface/server.cc#L553..L570

In other words, as long as, Thread B loads the boolean *after* adding something to the queue, we can't have that scenario.

That said, while writing this comment, it occurred to me that my argument only applies if there is only one thread A. This is completely broken if we have thread A1 and A2 doing the same steps you mentioned, back to back. A1 will see a non-empty queue, so it's safe. But A2, will see empty queues and process_pending_head is false.

To fix this I think I should use an atomic counter, that is incremented/decremented instead of boolean. We can only bypass the mutex if the counter is 0. Can you please double check this idea?

Reply:
The new upload has a bug. Please refrain from reviewing until I get that fixed. Thanks

Reply:
My suggestion is to tie the atomic counter to the size of the pending list. The current implementation (where it's not tied to the size) has a few problems:

1) Let's say Threads A1 and A2 reach line 1463 `gpr_atm_no_barrier_load(&rm->process_pending_head) == 0) ` and they are dealing with different cq indices. Let's say `&rm->process_pending_head` = 1. Then, A1 and A2 will both fall through and try to process the pending request. They will both eventually decrement `rm->process_pending_head` and it will be equal to -1. This is problematic because the next time something gets added to the pending list, the atomic will be updated to zero, and that request could get starved because queue_call_request will return early.

2) The other problem is that the atomic might get incremented in line 567, but a request may not get added to the pending list because we might return early around line 580. In queue_call_request, we might spuriously attempt to process the pending list, which kind of defeats the performance optimization. It probably won't happen often since we do double-check locking there, but it's a possibility.

One concern with this suggestion is that it could degrade performance with the extra atomic increments and decrements. You could probably get around this by keeping a local variable in the while loop at line 1469 to keep track of how many pending requests were handled and then just do one atomic decrement when the loop is finished.

Reply:
Thank you very much for double checking Hope! I thought about keeping the size, initially, but it suffers from the same problem of the original approach (PR #17401). When we keep the size, we basically want to bypass the lock when the size is 0. That is identical to checking whether the pending_head is `nullptr` as implemented in PR #17401.

In other words, we intentionally want to have the "problem 2" in your comment.

Now, thinking about this more I'm coming to a realization that we are probably trying too hard to optimize the existing approach. Maybe we should rethink how request matching is done in general. I'm going to refresh my mind on knots and bolts of this. I'd highly appreciate any suggestions/thoughts you have.

Reply:
Do you have new numbers from the latest commit?

Reply:
The latest commit looks fine to me. It solves problem 1 in my above comment, but still encounters problem 2. As long as the performance numbers show a win, this is okay.

Reply:
Thank you @hcaseyal ! (2) is intentional/required for fixing the problem on windows debug builds, but yes, I would not merge this patch unless we see some conclusive performance wins. I'll do that hopefully tomorrow.